### PR TITLE
Update consensus docs

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
Consensus docs are outdated

#### Summary of Changes
Update to latest consensus spec

Fixes #

### CODE REVIEW

File: docs/src/implemented-proposals/tower-bft.md [hunk 1]
```diff
@@ -16,25 +16,42 @@ For brevity this design assumes that a single voter with a stake is deployed as
...
+produce a block `B`. The `slot` of block `B` is designated `slot(B)`. Leaders
+do not necessarily need to generate a block for their slot, in which case there
+may not be blocks for some slots.
+
+For more details, see [fork generation](../cluster/fork-generation.md) and [leader rotation](../cluster/leader-rotation.md).
+
+## Votes
+
+Validators communicate which fork they think is the heaviest through votes.
+Each vote `v` is signed by the validator that produces it, and is of the form `(i, B)`, where `i` is the validaotr producing the vote and `B` is the block being voted for.
```
Comment (wen-coding) [inline docs/src/implemented-proposals/tower-bft.md:31]: nit: validaotr -> validator
Hmm, I think we have more than that in votes now?

File: docs/src/implemented-proposals/tower-bft.md [hunk 2]
```diff
@@ -16,25 +16,42 @@ For brevity this design assumes that a single voter with a stake is deployed as
...
+We call this stack the Vote Tower.
+
+When a vote is added to the tower, the lockouts of all the previous votes in the tower are doubled \(more on this in [Rollback](tower-bft.md#Rollback)\). With each new vote, a validator commits the previous votes to an ever-increasing lockout. At 32 votes we can consider the vote to be at `max lockout` any votes with a lockout equal to or above `1<<32` are dequeued \(FIFO\). Dequeuing a vote is the trigger for a reward. If a vote expires before it is dequeued, it and all the votes above it are popped \(LIFO\) from the vote tower. The validator needs to start rebuilding the tower from that point.
```
Comment (wen-coding) [inline docs/src/implemented-proposals/tower-bft.md:48]: nit: you changed Rolback to Vote Tower, probably should do a global replace.

File: docs/src/implemented-proposals/tower-bft.md [hunk 3]
```diff
@@ -67,21 +84,31 @@ At slot 10 the new votes caught up to the previous votes. But _vote 2_ expires a
...
 
-Each validator can independently set a threshold of cluster commitment to a fork before that validator commits to a fork. For example, at vote stack index 7, the lockout is 256 slots. A validator may withhold votes and let votes 0-7 expire unless the vote at index 7 has at greater than 50% commitment in the cluster. This allows each validator to independently control how much risk to commit to a fork. Committing to forks at a higher frequency would allow the validator to earn more rewards.
+In deciding whether to vote for a block `B`:
+
+1. Simulate a vote for `B` on your current tower
+2. Pop off all the votes that would be expired by `B`
+3. Now index every vote in the tower from `[0, tower.length()]`, assuming that the most recent simulated vote `B` is index 0, the second most recent vote is index 1, etc.
+4. Let `threshold_check_depth = 8` (the "depth" of 8 refers to this), now let `T` be the vote in the tower with index equal to threshold_check_depth.
+
+Now check all the blocks descended from `T`. If those blocks contain votes for `T` or any descendants of `T` totaling `>= 2/3` of the stake of the network, then we commit a vote to `T`.
```
Comment (wen-coding) [inline docs/src/implemented-proposals/tower-bft.md:111]: Shouldn't this line be 5. Now check ... ?

File: docs/src/implemented-proposals/tower-bft.md [hunk 4]
```diff
@@ -67,21 +84,31 @@ At slot 10 the new votes caught up to the previous votes. But _vote 2_ expires a
 |    7 |        11 |       2 |                   13 |
 |    1 |         1 |      16 |                   17 |
 
-The lockout for vote 1 will not increase from 16 until the stack contains 5 votes.
+The lockout for vote 1 will not increase from 16 until the tower contains 5 votes.
 
 ### Slashing and Rewards
 
-Validators should be rewarded for selecting the fork that the rest of the cluster selected as often as possible. This is well-aligned with generating a reward when the vote stack is full and the oldest vote needs to be dequeued. Thus a reward should be generated for each successful dequeue.
+Validators should be rewarded for selecting the fork that the rest of the cluster selected as often as possible. This is well-aligned with generating a reward when the vote tower is full and the oldest vote needs to be dequeued. Thus a reward should be generated for each successful dequeue.
```
[the remark this answers is not in the document]
Comment (carllin) [reply docs/src/implemented-proposals/tower-bft.md:91]: Removed this section

File: docs/src/implemented-proposals/tower-bft.md [hunk 5]
```diff
@@ -90,42 +117,47 @@ The following parameters need to be tuned:
...
 - Starting default lockout \(2\).
-- Threshold depth for minimum cluster commitment before committing to the fork \(8\).
+- Threshold check depth for minimum cluster commitment before committing to the fork \(8\).
 - Minimum cluster commitment size at threshold depth \(50%+\).
 
-### Free Choice
+### Fork Choice
 
-A "Free Choice" is an unenforcible validator action. There is no way for the protocol to encode and enforce these actions since each validator can modify the code and adjust the algorithm. A validator that maximizes self-reward over all possible futures should behave in such a way that the system is stable, and the local greedy choice should result in a greedy choice over all possible futures. A set of validator that are engaging in choices to disrupt the protocol should be bound by their stake weight to the denial of service. Two options exits for validator:
+Fork choice is how validators determine which fork to vote on given multiple
```
[the remark this answers is not in the document]
Comment (carllin) [reply docs/src/implemented-proposals/tower-bft.md:125]: clarified!

File: docs/src/implemented-proposals/tower-bft.md [hunk 6]
```diff
@@ -90,42 +117,47 @@ The following parameters need to be tuned:
...
+concurrent forks. Fork weights are based on the latest votes made by the validator set, and validators vote on the "heaviest" such fork.
```
[the remark this answers is not in the document]
Comment (carllin) [reply docs/src/implemented-proposals/tower-bft.md:126]: Switched to "Fork choice is how *each* validator determines" to specify it's a local algorithm. This is further emphasized in the algorithm which is in the context of: `Given the view of a single validator `i`:`

File: docs/src/implemented-proposals/tower-bft.md [hunk 7]
```diff
@@ -90,42 +117,47 @@ The following parameters need to be tuned:
...
-When evaluating multiple forks, each validator should use the following rules:
+1. For each vote `(j, B)` in `V`, add the stake of `j` to `B` and all of its
+ancestors.
+2. Now Set `B` to be the rooted block. Set `finish := 0`. 
 
-1. Forks must satisfy the _Threshold_ rule.
-2. Pick the fork that maximizes the total cluster lockout time for all the ancestor forks.
-3. Pick the fork that has the greatest amount of cluster transaction fees.
-4. Pick the latest fork in terms of PoH.
+While `finish == 0` do: If `i` has received no children of `B` then set `finish := 1` and return `B`. Otherwise let `B′` be the child of `B` (amongst those received by `i`) with most supporting votes in `V`, breaking ties by the smallest slot.
```
[the remark this answers is not in the document]
Comment (carllin) [reply docs/src/implemented-proposals/tower-bft.md:138]: Good suggestion, done!

File: docs/src/implemented-proposals/tower-bft.md (hunk 1 above)
Comment (carllin) [reply docs/src/implemented-proposals/tower-bft.md]: This is a high level description of the algorithm and so I simplified the vote to the minimum necessary to make the algorithm understandable.

If we want an exact documentation of the code, I think that should be done elsewhere

Comment (carllin) [reply docs/src/implemented-proposals/tower-bft.md]: Will update this to reflect that we include the whole tower once that feature is stabliized

File: docs/src/implemented-proposals/tower-bft.md [hunk 8]
```diff
@@ -67,21 +86,26 @@ At slot 10 the new votes caught up to the previous votes. But _vote 2_ expires a
...
 
 ### Cost of Rollback
 
 Cost of rollback of _fork A_ is defined as the cost in terms of lockout time to the validator to confirm any other fork that does not include _fork A_ as an ancestor.
 
 The **Economic Finality** of _fork A_ can be calculated as the loss of all the rewards from rollback of _fork A_ and its descendants, plus the opportunity cost of reward due to the exponentially growing lockout of the votes that have confirmed _fork A_.
 
-### Thresholds
+### Threshold Check
+In order to prevent a validator from locking themselves out on the wrong fork
```
Comment (wen-coding) [inline docs/src/implemented-proposals/tower-bft.md:98]: nit: locking itself

File: docs/src/implemented-proposals/tower-bft.md [hunk 9]
```diff
@@ -67,21 +86,26 @@ At slot 10 the new votes caught up to the previous votes. But _vote 2_ expires a
...
+in the case of a partition, there also needs to be a check to ensure that the rest of the cluster is committing to the same fork. This check is called the
+"threshold check", and is outlined as follows.
+
+In deciding whether to vote for a block `B`:
 
-Each validator can independently set a threshold of cluster commitment to a fork before that validator commits to a fork. For example, at vote stack index 7, the lockout is 256 slots. A validator may withhold votes and let votes 0-7 expire unless the vote at index 7 has at greater than 50% commitment in the cluster. This allows each validator to independently control how much risk to commit to a fork. Committing to forks at a higher frequency would allow the validator to earn more rewards.
+1. Simulate a vote for `B` on your current tower
+2. Pop off all the votes that would be expired by `B`
```
Comment (wen-coding) [inline docs/src/implemented-proposals/tower-bft.md:105]: This is a check, so it's not really pop off, it's also simulating pop off, right?

File: docs/src/implemented-proposals/tower-bft.md [hunk 10]
```diff
@@ -67,21 +86,26 @@ At slot 10 the new votes caught up to the previous votes. But _vote 2_ expires a
...
+3. Now index every vote in the tower from `[0, tower.length()]`, assuming that the most recent simulated vote `B` is index 0, the second most recent vote is index 1, etc.
+4. Let `threshold_check_depth = 8` (the "depth" of 8 refers to this), now let `T` be the vote in the tower with index equal to threshold_check_depth.
```
Comment (wen-coding) [inline docs/src/implemented-proposals/tower-bft.md:107]: Is thredshold_check_depth always 8?
Then you can change it to:
Let `T` be ... threshold_check_depth (currently hardcoded to be 8).

File: docs/src/implemented-proposals/tower-bft.md [hunk 11]
```diff
@@ -67,21 +86,26 @@ At slot 10 the new votes caught up to the previous votes. But _vote 2_ expires a
...
+5. Check all the blocks descended from `T`. If those blocks contain votes for `T` or any descendants of `T` totaling `>= 2/3` of the stake of the network, then we commit a vote to `T`.
```
Comment (wen-coding) [inline docs/src/implemented-proposals/tower-bft.md:108]: nit: In those blocks, if the validators voting for `T` or its descendants hold `>= 2/3` of the cluster's stake, then the check passes, we can commit a vote to `T`.

Not clear who holds the stake in the current sentence.

File: docs/src/implemented-proposals/tower-bft.md [hunk 12]
```diff
@@ -90,42 +114,49 @@ The following parameters need to be tuned:
...
+Fork choice is how each validator determines which fork to vote on when multiple
+concurrent forks exist. Forks are weighted are based on the latest votes made by the validator set, and individual validators then vote on the "heaviest"
+such fork.
 
-- a validator can outrun previous validator in virtual generation and submit a concurrent fork
-- a validator can withhold a vote to observe multiple forks before voting
+Given the view of a single validator `i`:
 
-In both cases, the validator in the cluster have several forks to pick from concurrently, even though each fork represents a different height. In both cases it is impossible for the protocol to detect if the validator behavior is intentional or not.
+Let `V` be the set of "most recent" valid votes received by `i`, i.e., `v = (j, B)` is in `V` and if `i` has not also received a vote of the form `(j, B′) `such that `slot(B′) > slot(B)`.
```
Comment (wen-coding) [inline docs/src/implemented-proposals/tower-bft.md:128]: nit: and if `i` -> and `i`

File: docs/src/implemented-proposals/tower-bft.md [hunk 13]
```diff
@@ -90,42 +114,49 @@ The following parameters need to be tuned:
...
+1. For each vote `(j, B)` in `V`, add the stake of `j` to `B` and all of its
+ancestors.
+2. Now Set `B` to be the rooted block. Set `finish := 0`.
 
-1. Forks must satisfy the _Threshold_ rule.
-2. Pick the fork that maximizes the total cluster lockout time for all the ancestor forks.
-3. Pick the fork that has the greatest amount of cluster transaction fees.
-4. Pick the latest fork in terms of PoH.
+*While `finish == 0`
+*do*: If `i` has received no children of `B` then set `finish := 1` and return `B`. Otherwise let `B′` be the child of `B` (amongst those received by `i`) with most the most stake-weighted votes in `V`, breaking ties by the smallest slot.
```
Comment (wen-coding) [inline docs/src/implemented-proposals/tower-bft.md:137]: nit: as a statement this is too long, would be better to use
if (...)
else
...
to put it on multiple lines.

File: docs/src/implemented-proposals/tower-bft.md [hunk 14]
```diff
@@ -16,25 +16,44 @@ For brevity this design assumes that a single voter with a stake is deployed as
 
 The Solana cluster generates a source of time via a Verifiable Delay Function we are calling [Proof of History](../cluster/synchronization.md).
 
-Proof of History is used to create a deterministic round robin schedule for all the active leaders. At any given time only 1 leader, which can be computed from the ledger itself, can propose a fork. For more details, see [fork generation](../cluster/fork-generation.md) and [leader rotation](../cluster/leader-rotation.md).
+Proof of History is used to create a deterministic round robin schedule for all the active leaders.
```
[the remark this answers is not in the document]
Comment (carllin) [reply docs/src/implemented-proposals/tower-bft.md:19]: agreed, removed!

File: docs/src/implemented-proposals/tower-bft.md [hunk 15]
```diff
@@ -16,25 +16,44 @@ For brevity this design assumes that a single voter with a stake is deployed as
...
+produce a block `B`. The `slot` of block `B` is designated `slot(B)`. Leaders
+do not necessarily need to generate a block for their slot, in which case there
+may not be blocks for some slots.
+
+For more details, see [fork generation](../cluster/fork-generation.md) and [leader rotation](../cluster/leader-rotation.md).
+
+## Votes
+
+Validators communicate which fork they think is the heaviest through votes.
+Each vote `v` is signed by the validator that produces it, and is of the form `(i, B)`, where `i` is the validator producing the vote and `B` is a hash
```
[the remark this answers is not in the document]
Comment (carllin) [reply docs/src/implemented-proposals/tower-bft.md:31]: yup, clarified!

File: docs/src/implemented-proposals/tower-bft.md (hunk 9 above)
Comment (carllin) [reply docs/src/implemented-proposals/tower-bft.md:105]: agreed, clarified

File: docs/src/implemented-proposals/tower-bft.md (hunk 11 above)
Comment (carllin) [reply docs/src/implemented-proposals/tower-bft.md:108]: Formalized this!

File: docs/src/implemented-proposals/tower-bft.md (hunk 13 above)
Comment (carllin) [reply docs/src/implemented-proposals/tower-bft.md:137]: Ah yeah, that's much better, thanks!

**Review by wen-coding — APPROVED** [review]

**Review by wen-coding — APPROVED** [review]

**Pull request merged**